Adding bedrock embedding

sriram-vinnakota opened the pull request:
Hi,

This is just initial work on bedrock.

changes to be noticed compared to previous azure embeddings:
**new param `isQuery`** : 
- Cohere had seperate way of embedding query and search doc.
- The embedded string of `isQuery = true` and `isQuery = false` are not same.
- but it amplify it's retrieval process.

**Embedding length:** 1024

**max token intake:** 512 (our previous model 8K)

**tokenization:** cohere did not made it's BPE based tokenization publicly available, so i used our old tokenizer to match the length

**Need to be done:**
- adding unit tests
- adding titan embedding models

Mael-Abgrall commented on src/ai/ai-bedrock-cohere-embedding.ts:
```diff
@@ -0,0 +1,155 @@
+      throw new Error('A document exceeds token limit');
+    }
+    const embeddingResponses = Promise.all(
+      processedChunks.map(async (chunk) => {
```
I'm worried this bit of the code could reach the rate limitation

Can you confirm:

- [ ] What is the rate limitation
- [ ] If the SDK help you avoid it (by limiting the number of requests, slowing them down)

If there is a rate limit and the SDK doesn't slow down your requests, you will need to slow them down yourself, AND remove the async map (because it will do it in parallel)

Mael-Abgrall commented on src/ai/ai-bedrock-cohere-embedding.ts:
```diff
@@ -0,0 +1,155 @@
+ * @param root.document the document to split into chunks
+ * @returns a list of chunks
+ */
+export async function createDocumentChunks({
```
Good improvement over my implementation of this function (I did not plan for filtering); However you still need to remove `\n` before the splitter (unless there is a reason to keep them of course)

Additionally, remove duplicate spaces, this can happen at times
```
// trim and remove duplicate whitespace
const trimmedDocument = cleanedDocument.replaceAll(/\s+/g, ' ').trim();
```

Mael-Abgrall commented on src/ai/ai-bedrock-cohere-embedding-model.ts:
```diff
@@ -0,0 +1,15 @@
+    | 'clustering';
+}
+
+export interface CohereEmbeddingResponse {
```
don't forget to link the doc whenever possible

Mael-Abgrall commented on src/ai/ai-bedrock-cohere-embedding-model.ts:
```diff
@@ -0,0 +1,15 @@
+export interface CohereEmbeddingResponse {
+  id: string;
+  texts: string[];
+  embeddings: number[][];
```
is there any explanation on why there might be multiple embeddings?

I know openAI has multiple embeddings when you send it multiple chunk at once, but it seems by your code here that you can only send one string per request, so having multiple embeddings as an output is odd

sriram-vinnakota replied:
@Mael-Abgrall, we can send multiple chunks at once but all of them have the same chunkID

Mael-Abgrall commented on src/ai/ai-bedrock-cohere-embedding-model.ts:
```diff
@@ -0,0 +1,15 @@
+export interface CohereEmbeddingBody {
+  texts: string[];
+  input_type:
```
any explanation here would be welcome; I saw your comment on the PR, but anyone seeing your code should know as well the use of `isQuery` and why it's important :)

recap, documenting is done like this:
```ts
/** documentation of the interface */ 
interface myInterface {
  /** documentation of the attribute */
  myAttribute: string;
}
```

Mael-Abgrall replied:
It also doesn't explain why there is classification and clustering

+ 

maybe for clarification I'd replace by a simpler `isQuery: boolean;` making it quite obvious when combined with the explanation


+ 

finally, if you are not using the interface elsewhere, maybe put it in the same document as the function, and don't export it; it reduce noise (I know this is not what we did until now, I'm learning as well)

Mael-Abgrall reviewed (commented):
Good work!

focus on solving the comments + integration tests (not necessary to have unit test yet).

btw, tests should have been done first, it would make it easier for you to develop too

do the titan models only once you are done with this please, don't split too much your attention

sriram-vinnakota commented:
Hi,

I removed preprocessing(new line and duplicate spaces) since the embedding does't change with or without preprocessing.
added sleep to avoid error for rate limit, need to add the link in the documentation [2000 calls per min ](https://docs.aws.amazon.com/general/latest/gr/bedrock.html)
